jax/experimental/rnn.py

```diff
@@ -308,12 +317,15 @@ def lstm_cell(carry, x, *, W_ih, W_hh, b_ih, b_hh):
   for l in range(num_layers * 2):
     cell = partial(
         lstm_cell, W_ih=W_ih[l], W_hh=W_hh[l], b_ih=b_ih[l], b_hh=b_hh[l])
+    cell_fn = partial(scan_fn, cell)
     if l % 2 == 0:
-      (h_t, c_t), seq_first_y_fwd = jax.lax.scan(cell, (h_0[l], c_0[l]),
+      out = jax.lax.scan(cell_fn, (h_0[l], c_0[l]),
                                                  seq_first_y)
+      (h_t, c_t), seq_first_y_fwd = _extract_output(seq_lengths, out)
     else:
-      (h_t, c_t), seq_first_y_bwd = jax.lax.scan(
-          cell, (h_0[l], c_0[l]), seq_first_y, reverse=True)
+      out = jax.lax.scan(
+          cell_fn, (h_0[l], c_0[l]), seq_first_y, reverse=True)
+      (h_t, c_t), seq_first_y_bwd = _extract_output(seq_lengths, out)
```

Comment:
`seq_lengths` need to be handled differently here since the sequences are reversed.

One solution would be to `jnp.roll` the sequences by `max_seq_len - seq_lengths`, then do the reversed scan, then roll them back by the same negative amount.

Reply:
This is tied to the previous comment, `_flip_sequence` should solve this.